provider/apple.go

```diff
@@ -0,0 +1,529 @@
+	if err != nil {
+		return errors.Wrap(err, "unmarshalling data from apple service response failed")
+	}
+
+	// If above operation done successfully checking a response code and error descriptions, if one exist.
+	// Apple service will response either 200 (OK) or 400 (any error).
+	if res.StatusCode != http.StatusOK || result.Error != "" {
+		return fmt.Errorf("apple token service error: %s", result.Error)
+	}
+
+	defer func() {
+		if err = res.Body.Close(); err != nil {
+			ah.L.Logf("[ERROR] close request body failed when get access token: %v", err)
+		}
+	}()
+
+	return err
+}
+
+// createClientSecret use for create the JWT client secret required to make requests to the Apple validation server.
+// for more details go to link: https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048
+func (ah *AppleHandler) createClientSecret() (string, error) {
+	if ah.conf.privateKey == nil {
+		return "", errors.New("private key can't be empty")
+	}
+	// Create the Claims
+	now := time.Now()
+	exp := now.Add(time.Second * 1800).Unix() // default value
+
+	if ah.conf.ClientSecretExpire != 0 {
+		exp = now.Add(time.Second * time.Duration(ah.conf.ClientSecretExpire)).Unix()
+	}
+
+	claims := &jwt.StandardClaims{
+		Issuer:    ah.conf.TeamID,
+		IssuedAt:  now.Unix(),
+		ExpiresAt: exp,
+		Audience:  "https://appleid.apple.com",
+		Subject:   ah.conf.ClientID,
+	}
+
+	tkn := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
+	tkn.Header["alg"] = "ES256"
+	tkn.Header["kid"] = ah.conf.KeyID
+
+	return tkn.SignedString(ah.conf.privateKey)
+}
+
+func (ah *AppleHandler) parseUserData(user *token.User, jUser string) {
+
+	type UserData struct {
+		Name struct {
+			FirstName string `json:"firstName"`
+			LastName  string `json:"lastName"`
+		} `json:"name"`
+		Email string `json:"email"`
+	}
+
+	var userData UserData
+
+	// Catch error for log only. No need break flow if user name doesn't exist
+	err := json.Unmarshal([]byte(jUser), &userData)
+	if err != nil {
+		ah.L.Logf("[ERROR] failed to parse user data %s: %v", user, err)
+		return
+	}
+	user.Name = fmt.Sprintf("%s %s", userData.Name.FirstName, userData.Name.LastName)
+}
+
+func (ah *AppleHandler) prepareLoginURL(state, path string) (string, error) {
+
+	scopesList := strings.Join(ah.conf.Scopes, " ")
+
+	authURL, err := url.Parse(ah.endpoint.AuthURL)
+	if err != nil {
+		return "", err
+	}
+
+	query := authURL.Query()
+	query.Set("state", state)
+	query.Set("response_type", "code")
+	query.Set("response_mode", "form_post")
+	query.Set("client_id", ah.conf.ClientID)
+	query.Set("scope", scopesList)
+	query.Set("redirect_uri", ah.makeRedirURL(path))
+	authURL.RawQuery = query.Encode()
+
+	return authURL.String(), nil
+
+}
+
+func (ah AppleHandler) makeRedirURL(path string) string {
+	elems := strings.Split(path, "/")
+	newPath := strings.Join(elems[:len(elems)-1], "/")
+
+	return strings.TrimRight(ah.URL, "/") + strings.TrimRight(newPath, "/") + urlCallbackSuffix
```

Comment:
TrimRight is not the same as TrimSuffix you probably meant

Reply:
I know you got it from oauth2, and it has the same issue. I'll fix oauth2 later on